File: src/services/studentCache.ts
/**
 * Student Cache Service
 * Provides persistent caching of RFID tag to student mappings for offline operation
 */

import { createLogger } from '../utils/logger';
import { safeInvoke } from '../utils/tauriContext';

import type { RfidScanResult } from './api';

const logger = createLogger('StudentCache');

// Cache entry structure
export interface CachedStudent {
  id: number;
  name: string;
  status: 'checked_in' | 'checked_out';
  lastSeen: string; // ISO timestamp
  room?: string;
  activity?: string;
  cachedAt: string; // ISO timestamp when cached
}

// Cache structure
export interface StudentCacheData {
  students: Record<string, CachedStudent>; // rfidTag -> CachedStudent
  metadata: {
    lastSync: string;
    version: number;
    dateCreated: string; // Date key for daily reset
  };
}

// Cache configuration
const CACHE_VERSION = 1;
const MAX_CACHE_AGE_HOURS = 24; // Cache expires after 24 hours

/**
 * Get today's date string for cache file naming
 */
function getTodayDateKey(): string {
  return new Date().toISOString().split('T')[0]; // Returns YYYY-MM-DD
}

// Removed getCacheFileName - not needed for frontend implementation

/**
 * Create empty cache structure
 */
function createEmptyCache(): StudentCacheData {
  return {
    students: {},
    metadata: {
      lastSync: new Date().toISOString(),
      version: CACHE_VERSION,
      dateCreated: getTodayDateKey(),
    },
  };
}

/**
 * Validate cache data structure and version
 */
function validateCacheData(data: unknown): data is StudentCacheData {
  if (!data || typeof data !== 'object') return false;
  const obj = data as Record<string, unknown>;
  if (!obj.students || typeof obj.students !== 'object') return false;
  if (!obj.metadata || typeof obj.metadata !== 'object') return false;
  const metadata = obj.metadata as Record<string, unknown>;
  if (typeof metadata.version !== 'number' || metadata.version !== CACHE_VERSION) return false;
  if (typeof metadata.dateCreated !== 'string' || metadata.dateCreated !== getTodayDateKey()) return false;
  return true;
}

/**
 * Check if cached entry is still fresh
 */
function isCacheEntryFresh(cachedAt: string): boolean {
  const cached = new Date(cachedAt);
  const now = new Date();
  const ageHours = (now.getTime() - cached.getTime()) / (1000 * 60 * 60);
  return ageHours <= MAX_CACHE_AGE_HOURS;
}

/**
 * Load student cache from persistent storage
 */
export async function loadStudentCache(): Promise<StudentCacheData> {
  try {
    logger.debug('Loading student cache from storage');

    const cacheData = await safeInvoke<StudentCacheData | null>('load_student_cache');

    if (!cacheData) {
      logger.info('No existing cache found, creating new cache');
      return createEmptyCache();
    }

    // Validate cache data
    if (!validateCacheData(cacheData)) {
      const obj = cacheData as Record<string, unknown>;
      const metadata = obj?.metadata as Record<string, unknown>;
      logger.warn('Invalid cache data found, creating new cache', {
        version: metadata?.version,
        dateCreated: metadata?.dateCreated,
        expectedDate: getTodayDateKey(),
      });
      return createEmptyCache();
    }

    // Clean up expired entries
    const cleanedStudents: Record<string, CachedStudent> = {};
    let expiredCount = 0;

    Object.entries(cacheData.students).forEach(([rfidTag, student]) => {
      if (isCacheEntryFresh(student.cachedAt)) {
        cleanedStudents[rfidTag] = student;
      } else {
        expiredCount++;
      }
    });

    const cleanedCache: StudentCacheData = {
      students: cleanedStudents,
      metadata: cacheData.metadata,
    };

    logger.info('Student cache loaded successfully', {
      totalEntries: Object.keys(cacheData.students).length,
      validEntries: Object.keys(cleanedStudents).length,
      expiredEntries: expiredCount,
      dateCreated: cacheData.metadata.dateCreated,
    });

    return cleanedCache;
  } catch (error) {
    logger.error('Failed to load student cache, creating new cache', { error });
    return createEmptyCache();
  }
}

/**
 * Save student cache to persistent storage
 */
export async function saveStudentCache(cacheData: StudentCacheData): Promise<void> {
  try {
    // Update metadata
    const updatedCache: StudentCacheData = {
      ...cacheData,
      metadata: {
        ...cacheData.metadata,
        lastSync: new Date().toISOString(),
      },
    };

    logger.debug('Saving student cache to storage', {
      entryCount: Object.keys(updatedCache.students).length,
    });

    await safeInvoke('save_student_cache', { settings: updatedCache });

    logger.info('Student cache saved successfully', {
      entryCount: Object.keys(updatedCache.students).length,
    });
  } catch (error) {
    logger.error('Failed to save student cache', { error });
    throw error;
  }
}

/**
 * Get cached student data for a specific RFID tag
 */
export function getCachedStudent(
  cache: StudentCacheData,
  rfidTag: string
): CachedStudent | null {
  const student = cache.students[rfidTag];

  if (!student) {
    logger.debug('No cached student found for tag', { rfidTag });
    return null;
  }

  if (!isCacheEntryFresh(student.cachedAt)) {
    logger.debug('Cached student data is expired', {
      rfidTag,
      cachedAt: student.cachedAt,
    });
    return null;
  }

  logger.debug('Retrieved cached student data', {
    rfidTag,
    studentId: student.id,
    studentName: student.name,
    status: student.status,
  });

  return student;
}

/**
 * Add or update student in cache
 */
export function setCachedStudent(
  cache: StudentCacheData,
  rfidTag: string,
  studentData: Omit<CachedStudent, 'cachedAt'>
): StudentCacheData {
  const now = new Date().toISOString();

  const updatedCache: StudentCacheData = {
    students: {
      ...cache.students,
      [rfidTag]: {
        ...studentData,
        cachedAt: now,
      },
    },
    metadata: cache.metadata,
  };

  logger.debug('Updated cached student data', {
    rfidTag,
    studentId: studentData.id,
    studentName: studentData.name,
    status: studentData.status,
  });

  return updatedCache;
}

/**
 * Convert API scan result to cached student format
 */
export function scanResultToCachedStudent(
  scanResult: RfidScanResult,
  additionalData?: {
    room?: string;
    activity?: string;
  }
): Omit<CachedStudent, 'cachedAt'> {
  return {
    id: scanResult.student_id,
    name: scanResult.student_name,
    status: scanResult.action === 'checked_out' ? 'checked_out' : 'checked_in',
    lastSeen: scanResult.processed_at ?? new Date().toISOString(),
    room: additionalData?.room ?? scanResult.room_name,
    activity: additionalData?.activity,
  };
}

/**
 * Remove expired entries from cache
 */
export function cleanExpiredEntries(cache: StudentCacheData): StudentCacheData {
  const cleanedStudents: Record<string, CachedStudent> = {};
  let expiredCount = 0;

  Object.entries(cache.students).forEach(([rfidTag, student]) => {
    if (isCacheEntryFresh(student.cachedAt)) {
      cleanedStudents[rfidTag] = student;
    } else {
      expiredCount++;
    }
  });

  if (expiredCount > 0) {
    logger.info('Cleaned expired cache entries', { expiredCount });
  }

  return {
    students: cleanedStudents,
    metadata: cache.metadata,
  };
}

/**
 * Get cache statistics
 */
export function getCacheStats(cache: StudentCacheData): {
  totalEntries: number;
  freshEntries: number;
  expiredEntries: number;
  checkedInCount: number;
  checkedOutCount: number;
  dateCreated: string;
  lastSync: string;
} {
  const entries = Object.values(cache.students);
  const freshEntries = entries.filter(s => isCacheEntryFresh(s.cachedAt));
  const expiredEntries = entries.length - freshEntries.length;

  return {
    totalEntries: entries.length,
    freshEntries: freshEntries.length,
    expiredEntries,
    checkedInCount: freshEntries.filter(s => s.status === 'checked_in').length,
    checkedOutCount: freshEntries.filter(s => s.status === 'checked_out').length,
    dateCreated: cache.metadata.dateCreated,
    lastSync: cache.metadata.lastSync,
  };
}

File: src-tauri/src/student_cache.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;
use tauri::{AppHandle, Manager};

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct CachedStudent {
    pub id: i32,
    pub name: String,
    pub status: String, // "checked_in" | "checked_out"
    #[serde(rename = "lastSeen")]
    pub last_seen: String, // ISO timestamp
    pub room: Option<String>,
    pub activity: Option<String>,
    #[serde(rename = "cachedAt")]
    pub cached_at: String, // ISO timestamp when cached
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct CacheMetadata {
    #[serde(rename = "lastSync")]
    pub last_sync: String,
    pub version: i32,
    #[serde(rename = "dateCreated")]
    pub date_created: String, // YYYY-MM-DD format
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct StudentCacheData {
    pub students: HashMap<String, CachedStudent>, // rfidTag -> CachedStudent
    pub metadata: CacheMetadata,
}

/// Get today's date string for cache file naming (YYYY-MM-DD)
fn get_today_date_key() -> String {
    chrono::Utc::now().format("%Y-%m-%d").to_string()
}

/// Generate cache filename for today
fn get_cache_filename() -> String {
    format!("student_cache_{}.json", get_today_date_key())
}

/// Get the path to the student cache file
fn get_student_cache_path(app_handle: &AppHandle) -> Result<PathBuf, String> {
    let app_data_dir = app_handle
        .path()
        .app_data_dir()
        .map_err(|e| format!("Failed to get app data directory: {}", e))?;

    // Ensure the directory exists
    fs::create_dir_all(&app_data_dir)
        .map_err(|e| format!("Failed to create app data directory: {}", e))?;

    Ok(app_data_dir.join(get_cache_filename()))
}

/// Validate cache data structure and check if it's for today
fn is_cache_valid(cache: &StudentCacheData) -> bool {
    // Check if cache is for today
    cache.metadata.date_created == get_today_date_key() && cache.metadata.version == 1
}

#[tauri::command]
pub async fn load_student_cache(app_handle: AppHandle) -> Result<Option<StudentCacheData>, String> {
    let cache_path = get_student_cache_path(&app_handle)?;

    // Check if file exists
    if !cache_path.exists() {
        return Ok(None);
    }

    let json_data = fs::read_to_string(&cache_path)
        .map_err(|e| format!("Failed to read student cache file: {}", e))?;

    let cache: StudentCacheData = serde_json::from_str(&json_data)
        .map_err(|e| format!("Failed to parse student cache: {}", e))?;

    // Validate cache (check if it's for today and has correct version)
    if !is_cache_valid(&cache) {
        // Cache is outdated or invalid, return None to create new cache
        return Ok(None);
    }

    Ok(Some(cache))
}

#[tauri::command]
pub async fn save_student_cache(
    app_handle: AppHandle,
    settings: StudentCacheData,
) -> Result<(), String> {
    let cache = settings;
    let cache_path = get_student_cache_path(&app_handle)?;

    // Update last sync timestamp
    let mut updated_cache = cache;
    updated_cache.metadata.last_sync = chrono::Utc::now().to_rfc3339();

    let json_data = serde_json::to_string_pretty(&updated_cache)
        .map_err(|e| format!("Failed to serialize student cache: {}", e))?;

    fs::write(&cache_path, json_data)
        .map_err(|e| format!("Failed to write student cache file: {}", e))?;

    Ok(())
}

#[tauri::command]
pub async fn clear_student_cache(app_handle: AppHandle) -> Result<(), String> {
    let cache_path = get_student_cache_path(&app_handle)?;

    // If file exists, remove it
    if cache_path.exists() {
        fs::remove_file(&cache_path)
            .map_err(|e| format!("Failed to remove student cache file: {}", e))?;
    }

    Ok(())
}

#[tauri::command]
pub async fn cleanup_old_student_caches(app_handle: AppHandle) -> Result<i32, String> {
    let app_data_dir = app_handle
        .path()
        .app_data_dir()
        .map_err(|e| format!("Failed to get app data directory: {}", e))?;

    let mut removed_count = 0;
    let today = get_today_date_key();

    // Read directory contents
    let entries = fs::read_dir(&app_data_dir)
        .map_err(|e| format!("Failed to read app data directory: {}", e))?;

    for entry in entries {
        let entry = entry.map_err(|e| format!("Failed to read directory entry: {}", e))?;
        let file_name = entry.file_name();
        let file_name_str = file_name.to_string_lossy();

        // Check if it's a student cache file and not today's
        if file_name_str.starts_with("student_cache_") && file_name_str.ends_with(".json") {
            // Extract date from filename
            if let Some(date_part) = file_name_str
                .strip_prefix("student_cache_")
                .and_then(|s| s.strip_suffix(".json"))
            {
                // If it's not today's cache, remove it
                if date_part != today {
                    if let Err(e) = fs::remove_file(entry.path()) {
                        eprintln!(
                            "Warning: Failed to remove old cache file {}: {}",
                            file_name_str, e
                        );
                    } else {
                        removed_count += 1;
                    }
                }
            }
        }
    }

    Ok(removed_count)
}

#[tauri::command]
pub async fn get_cache_stats(app_handle: AppHandle) -> Result<Option<serde_json::Value>, String> {
    let cache_path = get_student_cache_path(&app_handle)?;

    // Check if file exists
    if !cache_path.exists() {
        return Ok(None);
    }

    let json_data = fs::read_to_string(&cache_path)
        .map_err(|e| format!("Failed to read student cache file: {}", e))?;

    let cache: StudentCacheData = serde_json::from_str(&json_data)
        .map_err(|e| format!("Failed to parse student cache: {}", e))?;

    // Calculate stats
    let total_entries = cache.students.len();
    let checked_in_count = cache
        .students
        .values()
        .filter(|s| s.status == "checked_in")
        .count();
    let checked_out_count = cache
        .students
        .values()
        .filter(|s| s.status == "checked_out")
        .count();

    let stats = serde_json::json!({
        "total_entries": total_entries,
        "checked_in_count": checked_in_count,
        "checked_out_count": checked_out_count,
        "date_created": cache.metadata.date_created,
        "last_sync": cache.metadata.last_sync,
        "version": cache.metadata.version
    });

    Ok(Some(stats))
}
